### PR TITLE
Don't make any changes to the lhash structure if we are going to fail

### PR DESCRIPTION
**Pull request by mattcaswell**
The lhash expand() function can fail if realloc fails. The previous
implementation made changes to the structure and then attempted to do a
realloc. If the realloc failed then it attempted to undo the changes it
had just made. Unfortunately changes to lh->p were not undone correctly,
ultimately causing subsequent expand() calls to increment num_nodes to a
value higher than num_alloc_nodes, which can cause out-of-bounds reads/
writes. This is not considered a security issue because an attacker cannot
cause realloc to fail.

This commit moves the realloc call to near the beginning of the function
before any other changes are made to the lhash structure. That way if a
failure occurs we can immediately fail without having to undo anything.

Thanks to Pavel Kopyl (Samsung) for reporting this issue.

### CODE REVIEW

File: crypto/lhash/lhash.c [hunk 1]
```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
...
+    p = lh->p;
+    pmax = lh->pmax;
+    if (p + 1 >= pmax) {
+        j = nni * 2;
+        n = OPENSSL_realloc(lh->b, sizeof(OPENSSL_LH_NODE *) * j);
+        if (n == NULL) {
+            lh->error++;
+            return 0;
+        }
+        for (i = nni; i < j; i++)
```
Comment (richsalz) [inline crypto/lhash/lhash.c]: Make this a single memset call.  Then you can get rid of either nni or j variables.

File: crypto/lhash/lhash.c [hunk 2]
```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
...
+            n[i] = NULL;
+        lh->pmax = lh->num_alloc_nodes;
+        lh->num_alloc_nodes = j;
+        lh->num_expand_reallocs++;
+        lh->p = 0;
+        lh->b = n;
```
Comment (richsalz) [inline crypto/lhash/lhash.c]: Move this ```lh->b``` assignment right up after the NULL check for n.

File: crypto/lhash/lhash.c [hunk 3]
```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
 static int expand(OPENSSL_LHASH *lh)
 {
     OPENSSL_LH_NODE **n, **n1, **n2, *np;
-    unsigned int p, i, j;
-    unsigned long hash, nni;
+    unsigned int p, pmax, nni, i, j;
+    unsigned long hash;
+
+    nni = lh->num_alloc_nodes;
+    p = lh->p;
```
Comment (richsalz) [inline crypto/lhash/lhash.c]: I don't see that ```p``` is used.

File: crypto/lhash/lhash.c (hunk 1 above)
Comment (vdukhovni) [reply crypto/lhash/lhash.c]: @richsalz I see we're continuing to assume/require that NULL is zero-filled :-)

File: crypto/lhash/lhash.c (hunk 3 above)
Comment (vdukhovni) [reply crypto/lhash/lhash.c]: It is used.

File: crypto/lhash/lhash.c [hunk 4]
```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
...
+    } else {
+        lh->p++;
+    }
 
     lh->num_nodes++;
     lh->num_expands++;
-    p = (int)lh->p++;
     n1 = &(lh->b[p]);
-    n2 = &(lh->b[p + (int)lh->pmax]);
+    n2 = &(lh->b[p + pmax]);
```
Comment (vdukhovni) [inline crypto/lhash/lhash.c]: Why are we zeroing `b[p + pmax]`.  This function needs to be rewritten in deobfuscated modern C, with fewer cryptic variables and with some comments to explain mystery arithmetic such as the above.

File: crypto/lhash/lhash.c [hunk 5]
```diff
@@ -186,16 +186,35 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
...
+        }
+        for (i = nni; i < j; i++)
+            n[i] = NULL;
+        lh->pmax = lh->num_alloc_nodes;
+        lh->num_alloc_nodes = j;
+        lh->num_expand_reallocs++;
+        lh->p = 0;
+        lh->b = n;
+    } else {
+        lh->p++;
```
Comment (vdukhovni) [inline crypto/lhash/lhash.c]: I am rather perplexed as to why we set `lh->p = 0` when reallocating, and increment it instead otherwise?  Should it not end up at the same value either way?  Why should the resulting state be different if we needed a realloc than when we did not?  What is `lh->p` for?

Comment (mattcaswell) [reply crypto/lhash/lhash.c]: It's a great question. After over an hour of studying this code I still could not figure out exactly what `p` is used for or why it works the way it does. It appears to be some kind of counter that increments each time we add nodes. As soon as `p` hits `pmax` we add more nodes, update `pmax` and reset the counter `p`. That's why things are different in the realloc case, i.e. we are resetting the counter, whereas otherwise we just increment it.

In the end I opted for maintaining the current behaviour (this is actually what the current code does - it increments `p` unconditionally, and then resets it to 0 in the event of a realloc). This seemed the simplest solution.

File: crypto/lhash/lhash.c (hunk 4 above)
Comment (mattcaswell) [reply crypto/lhash/lhash.c]: I completely agree and considered that myself. But, frankly, a complete rewrite would require more attention from me that I am able to give it at the moment. Happy for someone else to pick it up though. In any case a complete rewrite is probably only appropriate for master - we still need to fix 1.1.0 and 1.0.2.

File: crypto/lhash/lhash.c (hunk 5 above)
Comment (vdukhovni) [reply crypto/lhash/lhash.c]: OK, it seems that `p` is the first index for which the hashes may need to be recomputed after the hash table expands.  If a number is congruent to `x mod n` then it is either `x` or `x + n` mod `2n` so the hash table update moves some nodes from `x` to `x + n` where `n` is `pmax` and the table size is `2 * pmax`.

There are way too many redundant poorly documented variables here.  This code needs to be put on the urgent need for repair list.

Comment (vdukhovni) [reply crypto/lhash/lhash.c]: Hash nodes with  `hash < p mod pmax` have up-to-date hash slots based on the most recent expansion of the table.  The others are hashed with the previous modulus.  As we call expand repeatedly, eventually `p == pmax` and all the hashes become current, at which point we actually grow the array and invalidate all the hashes relative to the new modulus (so they all continue to use the old modulus).

File: crypto/lhash/lhash.c (hunk 1 above)
Comment (richsalz) [reply crypto/lhash/lhash.c]: Yes.  And in fact we test for it, see test/sanitytest.c

**Comment by mattcaswell** [comment]
New commit pushed addressing feedback from @richsalz. I've not updated the other one yet - I'll wait until I get approval for this first.

File: crypto/lhash/lhash.c [hunk 6]
```diff
@@ -199,13 +199,12 @@ static int expand(OPENSSL_LHASH *lh)
             lh->error++;
             return 0;
         }
-        for (i = nni; i < j; i++)
-            n[i] = NULL;
+        lh->b = n;
+        memset(n + nni, 0, sizeof(*n) * nni);
```
Comment (vdukhovni) [inline crypto/lhash/lhash.c]: For clarity replace `sizeof(*n) * nni`  with `sizeof(*n) * (j -nni)`

**Comment by mattcaswell** [comment]
Updated commit pushed addressing @vdukhovni's comment

**Comment by vdukhovni** [comment]
I'd like to see comments in the file documenting the origin of the algorithm.  With the help of Thor Simon the following were found:

* Litwin, Witold (1980), "Linear hashing: A new tool for file and table addressing" (PDF), Proc. 6th Conference on Very Large Databases: 212–223 <http://www.cs.cmu.edu/afs/cs.cmu.edu/user/christos/www/courses/826-resources/PAPERS+BOOK/linear-hashing.PDF>
* More related info <https://en.wikipedia.org/wiki/Linear_hashing#Adoption_in_database_systems>

Specifically:

> Linear hashing is used in the BDB Berkeley database system, which in turn is used by many software systems such as OpenLDAP, using a C implementation derived from the CACM article and first published on the Usenet in 1988 by Esmond Pitt.

I found a non-paywalled copy of the full CACM article at <https://pdfs.semanticscholar.org/ff4d/1c5deca6269cc316bfd952172284dbf610ee.pdf>

And I'd still like to see a less obscure coding style in this file...

**Comment by paulidale** [comment]
I agree, this code is hard to follow.  It took me way to long to figure out how it worked.  I've got _rewrite lhash using more modern techniques_ on my to do list but I've no idea when or if I'll get to it.

I thought there was a reference to the paper used in the code (but I can't find it).  Lhash implements size doublings which isn't the idea behind linear hashes.  The slot reallocations are amortised like a linear hash would.  Given the power of two size and the test a bit reallocation (which isn't used), the reallocation could be simplified.

Does anyone have any idea as to the growth of lhash tables?  More particularly, about their shrinkage?  Does the delete operation need to downsize the array?  There aren't a lot of deletes in the code.

File: crypto/lhash/lhash.c [hunk 7]
```diff
@@ -186,16 +186,34 @@ void OPENSSL_LH_doall_arg(OPENSSL_LHASH *lh, OPENSSL_LH_DOALL_FUNCARG func, void
...
+    if (p + 1 >= pmax) {
+        j = nni * 2;
+        n = OPENSSL_realloc(lh->b, sizeof(OPENSSL_LH_NODE *) * j);
+        if (n == NULL) {
+            lh->error++;
+            return 0;
+        }
+        lh->b = n;
+        memset(n + nni, 0, sizeof(*n) * (j - nni));
+        lh->pmax = lh->num_alloc_nodes;
```
Comment (bernd-edlinger) [inline crypto/lhash/lhash.c]: you could use "nni" instead of "lh->num_alloc_nodes" here?

**Comment by mattcaswell** [comment]
I added some new commits (all to be squashed) addressing the request from @vdukhovni for a comment documenting the origin of the algorithm, and another commit addressing @bernd-edlinger comment.

**Comment by mattcaswell** [comment]
Pushed. Thanks. I will update the 1.0.2 version of this PR to match the changes made here shortly.
